pkg/serviceaccount/metadata/metadata.go

```diff
@@ -0,0 +1,217 @@
+func fromStandard(h http.HandlerFunc) restful.RouteFunction {
+	return func(req *restful.Request, resp *restful.Response) {
+		h(resp, req.Request)
+	}
+}
+
+func (s *IssuerMetadataServer) serveConfiguration(w http.ResponseWriter, req *http.Request) {
+	b, err := json.Marshal(s.metadata)
+	if err != nil {
+		klog.Errorf("failed to marshal service account issuer metadata: %v", err)
+		http.Error(w, "internal error", http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Header().Set("Cache-Control", "max-age=1")
+	if _, err := w.Write(b); err != nil {
+		klog.Errorf("failed to write service account issuer metadata response: %v", err)
+		return
+	}
+}
+
+func getAlgs(keys *jose.JSONWebKeySet) []string {
+	var result []string
+	algs := map[string]bool{}
+	for _, k := range keys.Keys {
+		a := k.Algorithm
+		if !algs[a] {
+			result = append(result, a)
+		}
+		algs[a] = true
+	}
+	sort.Strings(result)
+	return result
+}
+
+func getJwks(in []interface{}) (*jose.JSONWebKeySet, errors.Aggregate) {
+	// Decode keys into a JWKS.
+	var keys jose.JSONWebKeySet
+	var errs []error
+	for i, key := range in {
+		var pubkey *jose.JSONWebKey
+		var err error
+
+		switch k := key.(type) {
+		case interface {
+			Public() crypto.PublicKey
+		}:
+			// This is a private key. Get its public key
+			pubkey, err = jwkFromPubkey(k.Public())
+		case rsa.PublicKey:
+			pubkey, err = jwkFromPubkey(&k)
+		case *rsa.PublicKey:
+			pubkey, err = jwkFromPubkey(k)
+		case ecdsa.PublicKey:
+			pubkey, err = jwkFromPubkey(&k)
+		case *ecdsa.PublicKey:
+			pubkey, err = jwkFromPubkey(k)
+		default:
+			err = fmt.Errorf("key must be rsa.PublicKey or ecdsa.PublicKey")
+		}
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if !pubkey.Valid() {
+			errs = append(errs, fmt.Errorf("configured key #%d not valid", i))
+			continue
+		}
+		keys.Keys = append(keys.Keys, *pubkey)
+	}
+	if len(errs) != 0 {
+		return &keys, errors.NewAggregate(errs)
+	}
+	return &keys, nil
+}
+
+func (s *IssuerMetadataServer) serveKeys(w http.ResponseWriter, req *http.Request) {
+	b, err := json.Marshal(s.keys)
+	if err != nil {
+		klog.Errorf("failed to marshal service account issuer JWKS: %v", err)
+		http.Error(w, "internal error", http.StatusInternalServerError)
+		return
+	}
+
+	// Per RFC7517 : https://tools.ietf.org/html/rfc7517#section-8.5.1
+	w.Header().Set("Content-Type", "application/jwk-set+json")
+	w.Header().Set("Cache-Control", "max-age=1")
+	if _, err := w.Write(b); err != nil {
+		klog.Errorf("failed to write service account issuer JWKS response: %v", err)
+		return
+	}
+}
+
+func jwkFromPubkey(k crypto.PublicKey) (*jose.JSONWebKey, error) {
+	alg, err := serviceaccount.AlgorithmForPublicKey(k)
+	if err != nil {
+		return nil, fmt.Errorf("must be rsa.PublicKey, ecdsa.PublicKey, or jose.OpaqueSigner")
+	}
+	return &jose.JSONWebKey{
```

Comment:
Hmm, looks like we should be setting a key ID here. I suspect this is why the e2e test is currently failing on my local setup.

Reply:
Updated to set key id, and hey now the e2e test passes!